core/src/main/scala/org/http4s/HttpVersion.scala

```diff
@@ -35,20 +35,21 @@ object HttpVersion {
       case "HTTP/1.1" => Either.right(`HTTP/1.1`)
       case "HTTP/1.0" => Either.right(`HTTP/1.0`)
       case _ =>
-        new Parser(s).HttpVersion.run()(Parser.DeliveryScheme.Either).leftMap { _ =>
-          ParseFailure("Invalid HTTP version", s"$s was not found to be a valid HTTP version")
+        parser.parseAll(s).leftMap { case e =>
+          ParseFailure("Invalid HTTP version", e.toString)
         }
     }
 
-  private class Parser(val input: ParserInput)
-      extends org.http4s.internal.parboiled2.Parser
-      with Rfc2616BasicRules {
-    def HttpVersion: Rule1[org.http4s.HttpVersion] =
-      rule {
-        "HTTP/" ~ capture(Digit) ~ "." ~ capture(Digit) ~> { (major: String, minor: String) =>
-          new HttpVersion(major.toInt, minor.toInt)
-        }
-      }
+  private val parser: Parser1[HttpVersion] = {
```

Comment:
Some of our parsers have existed on the companion of what they're parsing so they can access a private constructor after validation.  It prevents a `private [http4s] unsafeFromVersion(Int, Int)`.

Do we like this pattern or no?  The lack of inheritance perhaps makes it more palatable than it was.

Reply:
I use this pattern in internal libs a lot, and have found it quite useful.

Reply:
As long as we have no circular dependencies between parsed types, I like it, too.  No way to construct an invalid one outside this compilation unit.